integration/spark/app/src/main/java/io/openlineage/spark/agent/OpenLineageSparkListener.java

```diff
@@ -264,30 +245,99 @@ public static void close() {
    */
   @Override
   public void onApplicationStart(SparkListenerApplicationStart applicationStart) {
-    initializeContextFactoryIfNotInitialized();
+    initializeContextFactoryIfNotInitialized(applicationStart.appName());
+    emitApplicationStartEvent(applicationStart.time());
   }
 
   private void initializeContextFactoryIfNotInitialized() {
+    if (contextFactory != null || isDisabled) {
+      return;
+    }
+    asJavaOptional(activeSparkContext.apply())
+        .ifPresent(context -> initializeContextFactoryIfNotInitialized(context.appName()));
+  }
+
+  private void initializeContextFactoryIfNotInitialized(String appName) {
     if (contextFactory != null || isDisabled) {
       return;
     }
     SparkEnv sparkEnv = SparkEnv$.MODULE$.get();
-    if (sparkEnv != null) {
-      try {
-        ArgumentParser args = ArgumentParser.parse(sparkEnv.conf());
-        contextFactory = new ContextFactory(new EventEmitter(args));
-      } catch (URISyntaxException e) {
-        log.error("Unable to parse open lineage endpoint. Lineage events will not be collected", e);
-      }
-    } else {
+    if (sparkEnv == null) {
       log.warn(
           "Open lineage listener instantiated, but no configuration could be found. "
               + "Lineage events will not be collected");
+      return;
+    }
+    initializeContextFactoryIfNotInitialized(sparkEnv.conf(), appName);
+  }
+
+  private void initializeContextFactoryIfNotInitialized(SparkConf sparkConf, String appName) {
+    if (contextFactory != null || isDisabled) {
+      return;
+    }
+    try {
+      ArgumentParser args = ArgumentParser.parse(sparkConf);
+      contextFactory = new ContextFactory(new EventEmitter(args, appName));
+    } catch (URISyntaxException e) {
+      log.error("Unable to parse open lineage endpoint. Lineage events will not be collected", e);
     }
   }
 
+  public void emitApplicationEvent(Long time, OpenLineage.RunEvent.EventType eventType) {
+    OpenLineage openLineage = new OpenLineage(Versions.OPEN_LINEAGE_PRODUCER_URI);
+    EventEmitter emitter = contextFactory.openLineageEventEmitter;
+    OpenLineage.RunBuilder runBuilder =
+        openLineage.newRunBuilder().runId(emitter.getApplicationRunId());
+    if (emitter.getParentRunId().isPresent()
+        && emitter.getParentJobName().isPresent()
+        && emitter.getParentJobNamespace().isPresent()) {
+      runBuilder.facets(
+          openLineage
+              .newRunFacetsBuilder()
+              .parent(
+                  openLineage.newParentRunFacet(
+                      openLineage.newParentRunFacetRun(emitter.getParentRunId().get()),
+                      openLineage.newParentRunFacetJob(
+                          emitter.getParentJobNamespace().get(), emitter.getParentJobName().get())))
+              .build());
+    }
+    OpenLineage.RunEvent applicationEvent =
```

Comment:
wouldn't it be nice to unit test this method?

Reply:
Added unit tests for onApplicationStart and onApplicationEnd methods.